File: util-java/src/main/java/ch/bailu/util_java/parser/scanner/SimpleStringReader.java
package ch.bailu.util_java.parser.scanner;

import java.io.IOException;
import java.io.Reader;

public class SimpleStringReader extends Reader {

    private String string = new String();
    private int index = 0;

    public void setString(String s) throws IOException {
        string = s;
        reset();
    }


    @Override
    public int read() throws IOException {
        if (index < string.length()) {
            int c = string.charAt(index);
            index ++;
            return c;
        }
        return -1;
    }


    @Override
    public int read(char[] cbuf) throws IOException {
        return read(cbuf, 0, cbuf.length);
    }



    @Override
    public long skip(long n) throws IOException {
        if (index + n >= string.length()) {
            n = string.length()- index;
        }
        index += n;

        return n;
    }

    @Override
    public boolean ready() throws IOException {
        return true;
    }


    @Override
    public void reset() throws IOException {
        index = 0;
    }





    @Override
    public int read(char[] cbuf, int off, int len) throws IOException {
        if (index >= string.length()) return -1;

        int c = 0;
        int out = off;

        while (c < len && out < cbuf.length ) {
            cbuf[off] = string.charAt(index);

            if (index + 1 < string.length()) {
                index ++;
                c++;
                out++;
            } else {
                break;
            }
        }
        return c;
    }


    @Override
    public void close() throws IOException {

    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/xml/parser/util/SimpleStringReader.kt
package ch.bailu.aat_lib.xml.parser.util

import java.io.IOException
import java.io.Reader

class SimpleStringReader : Reader() {
    private var string = ""
    private var index = 0

    @Throws(IOException::class)
    fun setString(s: String) {
        string = s
        reset()
    }

    override fun read(): Int {
        if (index < string.length) {
            val c = string[index].code
            index++
            return c
        }
        return -1
    }

    @Throws(IOException::class)
    override fun read(cbuf: CharArray): Int {
        return read(cbuf, 0, cbuf.size)
    }

    override fun skip(n: Long): Long {
        var result = n
        if (index + result >= string.length) {
            result = (string.length - index).toLong()
        }
        index += result.toInt()
        return result
    }

    override fun ready(): Boolean {
        return true
    }

    override fun reset() {
        index = 0
    }

    override fun read(cbuf: CharArray, off: Int, len: Int): Int {
        if (index >= string.length) return -1
        var c = 0
        var out = off
        while (c < len && out < cbuf.size) {
            cbuf[off] = string[index]
            if (index + 1 < string.length) {
                index++
                c++
                out++
            } else {
                break
            }
        }
        return c
    }

    override fun close() {}
}
